### PR TITLE
map: Transfer to MapLibre

### LINKED ISSUE #26739: [$300 bounty] nav: switch to MapLibre widget
Seems like it's being actively developed now: https://github.com/maplibre/maplibre-gl-native.

Bounty is for making the simplest PR that moves to MapLibre. Requirements:
* performance must be same or better
* map must look the same (or very good reason if not)

This can be done on PC, and we can do final testing on a comma 3X.

### PR DESCRIPTION
**Pull request by bongbui321**
**Description**
Currently translate the codebase to maplibre-qt already. 

**Verification**
Trying to build maplibre-qt, but is currently facing an issue similar to [this](https://github.com/maplibre/maplibre-native-qt/issues/76). Failing to build because of missing `qtlocation5-private-dev` package and it is not available on apt

Will  test using `map_renderer.py`, not sure if this is a good way to test this? Probably better test with replay

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
> Will test using `map_renderer.py`, not sure if this is a good way to test this? Probably better test with replay

I'd probably start with replaying the demo route and just checking the map generally works in the UI, then test the map renderer. See here for getting setup: https://github.com/commaai/openpilot/tree/master/selfdrive/navd#development

**Comment by adeebshihadeh** [comment]
What does full desktop support mean? We can't make this work with 5.12?

**Comment by bongbui321** [comment]
From the [README](https://github.com/maplibre/maplibre-native-qt), it is "fully supported on desktop platform" which I think on the OSes. I think it is fine since Agnos is mainly based on Ubuntu 20.04? 

anything below 5.15 doesn't support `Location` which I think is used for getting routes and positions. I think we can kind of work around this if we have our own fork. I haven't dive deep into the codebase, but it might be possible

**Comment by adeebshihadeh** [comment]
Ah, if there's lot of hacks, then I don't think this is worth it. We'll be switching to 24.04 in a few months when it's out anyway. If you confirm, there are lots of hacks to get it working on our Qt, then we can just keep this bounty locked to you until the switch.

**Comment by bongbui321** [comment]
Okay, I think it is possible to work with 5.12. I made a small [PR](https://github.com/maplibre/maplibre-native-qt/pull/107) that makes it compilable with Qt 5.12. We don't need `Location` since we are getting position and location from cereal already. so we can disable the `Location` option with 
`cmake -DMLN_QT_WITH_LOCATION=OFF ..`

I was trying to test with the demo route of `replay` but that one doesn't seem to have map enabled, is there a different route that I can test with?. 
I also test with `map_renderer.py` and got a `segfault`. I tried to implement the exact same function in `.c` here is the file: 
 [test.txt](https://github.com/commaai/openpilot/files/14075441/test.txt) and it works just fine with no `segfault`, not entirely sure why it doesn't work for python implementation

**Comment by adeebshihadeh** [comment]
> Okay, I made it work with 5.12. I made a small [PR](https://github.com/maplibre/maplibre-native-qt/pull/107) that makes it compilable with Qt 5.12. We don't need `Location` since we are getting position and location from cereal already. so we can disable the `Location` option with `cmake -DMLN_QT_WITH_LOCATION=OFF ..`
> 
> I was trying to test with the demo route of `replay` but that one doesn't seem to have map enabled, is there a different route that I can test with?. I also test with `map_renderer.py` and got a `segfault`. I tried to implement the exact same function in `.c` here is the file: [test.txt](https://github.com/commaai/openpilot/files/14075441/test.txt) and it works just fine with no `segfault`, not entirely sure why it doesn't work for python implementation

Nice! I see that got merged already. 

Testing it doesn't require a route with nav enabled; the map will be available if you have everything setup, such as a token.

**Comment by bongbui321** [comment]
@adeebshihadeh it seems to work perfectly now, do you need me to upload the compiled library or do you want me to write a shell to download and set it up automatically. Here is a snapshot of the demo with maplibre. Still figuring out how to record videos :)
image

**Comment by adeebshihadeh** [comment]
Great. We always want a script to build the `third_party` dependencies so that they're reproducible, then you can check them in once that's done. This one is a good example https://github.com/commaai/openpilot/blob/master/third_party/acados/build.sh

**Comment by adeebshihadeh** [comment]
@mitchellgoffpc any validation you want to do before merging this?

**Comment by bongbui321** [comment]
@adeebshihadeh added the build script

> I also test with map_renderer.py and got a segfault. I tried to implement the exact same function in .c here is the file: [test.txt](https://github.com/commaai/openpilot/files/14075441/test.txt) and it works just fine with no segfault, not entirely sure why it doesn't work for python implementation

The UI works fine, if possible you can test it out just to make sure I didn't introduce any bug, but the `map_renderer.py` doesn't seem to work. I created the same code in `.c` file as above and it doesn't have any segfault, I'm not sure why it doesn't work

**Comment by mitchellgoffpc** [comment]
> @mitchellgoffpc any validation you want to do before merging this?

Yeah we should verify that it doesn't impact model performance, will run those tests tomorrow morning.

File: third_party/maplibre-native-qt/build.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  ARCHNAME="Darwin"
+fi
+
+if [ ! -d maplibre_repo ]; then
+  git clone git@github.com:maplibre/maplibre-native-qt.git $DIR/maplibre_repo
+fi
+cd maplibre_repo
+git fetch --all
+# git checkout (on main right now)
+git submodule update --depth=1 --recursive --init
```
Comment (adeebshihadeh) [inline third_party/maplibre-native-qt/build.sh:17]: this should be pinned to a specific commit on their repo. in this case, likely the commit where they merged your fix

File: third_party/maplibre-native-qt/build.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#build
+mkdir -p build && cd build
```
Comment (adeebshihadeh) [inline third_party/maplibre-native-qt/build.sh:20]:
```suggestion
mkdir -p build
cd build
```

File: third_party/maplibre-native-qt/build.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+if [ ! -d maplibre_repo ]; then
+  git clone git@github.com:maplibre/maplibre-native-qt.git $DIR/maplibre_repo
+fi
+cd maplibre_repo
+git fetch --all
+# git checkout (on main right now)
+git submodule update --depth=1 --recursive --init
+
+#build
```
Comment (adeebshihadeh) [inline third_party/maplibre-native-qt/build.sh:19]:
```suggestion
# build
```

File: third_party/maplibre-native-qt/build.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+set -e
+
+DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null && pwd)"
+
+ARCHNAME="x86_64"
```
Comment (adeebshihadeh) [inline third_party/maplibre-native-qt/build.sh:6]: can you add support for aarch64 for the comma 3X? I can handle testing it

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
I haven't run it yet, but it looks pretty good! I left a few minor things to cleanup.

**Comment by adeebshihadeh** [comment]
CI is passing now, but there's some issues building it on device (CMake is too old and it's missing `libqconnmanbearer.so`). I'll look into those when I get a chance.

**Comment by adeebshihadeh** [comment]
Got it working on the comma 3X, and all the tests pass except navModel replay. I'll update the refs once you verify @mitchellgoffpc.

**Comment by mitchellgoffpc** [comment]
@adeebshihadeh lgtm, model behavior appears to be pretty much identical

**Comment by SurferSD** [comment]
This merge (transfer to MapLibre) breaks OP tools build (PJ plugin) on the Apple platform due to missing Darwin library files.

**Comment by bongbui321** [comment]
@SurferSD please use the build script, and i think its better to open an issue rather than commenting on Pr since it mostly relates to development

**Pull request merged**